scheduler perf: benchstat support

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Makes it possible to compare the performance impact of a code change without relying on the performance dashboard.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

Example output after
```
go test -timeout=0 -count=3 ./test/integration/scheduler_perf/ -short=true -run=none -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic >/tmp/log 2>&1
benchstat /tmp/log
```

```
name                                         SchedulingThroughput/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 500 ± 0%

name                                         SchedulingThroughput/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 375 ± 3%

name                                         SchedulingThroughput/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         SchedulingThroughput/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         SchedulingThroughput/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         scheduler_framework_extension_point_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.58 ± 2%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.00 ± 3%

name                                         scheduler_framework_extension_point_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.56 ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.04 ± 2%

name                                         scheduler_framework_extension_point_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.77 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.51 ± 0%

name                                         scheduler_framework_extension_point_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.81 ± 2%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.57 ± 0%

name                                         scheduler_framework_extension_point_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                 2.19 ± 5%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                2.97 ± 9%

name                                         scheduler_pod_scheduling_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                  375 ± 4%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 700 ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                  348 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 687 ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                  956 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               1.61k ± 5%

name                                         scheduler_pod_scheduling_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                1.12k ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               2.09k ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                1.25k ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               2.47k ± 0%

name                                         scheduler_scheduling_attempt_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                 3.48 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                3.97 ± 5%

name                                         scheduler_scheduling_attempt_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                 3.08 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                3.76 ± 4%

name                                         scheduler_scheduling_attempt_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                 5.35 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                7.30 ± 2%

name                                         scheduler_scheduling_attempt_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                 7.26 ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                7.80 ± 1%

name                                         scheduler_scheduling_attempt_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                 14.6 ± 4%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                16.2 ±27%
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @kerthcet

pohly committed:
scheduler_perf: report data items as benchmark results

This replaces the pretty useless us/op metric (useless because it includes
setup and teardown times) with the same values that also get stored in the JSON
file.

The main advantage is that benchstat can be used to analyze and compare
results.

pohly committed:
test/integration: add StartEtcd

In contrast to EtcdMain, it can be called by individual tests or benchmarks and
each caller will get a fresh etcd instance. However, it uses the same
underlying code and the same port for all instances, so tests cannot run in
parallel.

kerthcet commented:
/retest

pohly committed:
test/utils: extend ktesting

The upstream ktesting has to be very flexible to accommodate different ways of
using it. In Kubernetes, we can be opinionated and make certain choices, like
using klog flags, and only those.

kerthcet commented:
Can this one be independent, or you want to wait until https://github.com/kubernetes/kubernetes/pull/115422 merged.

pohly commented:
The other PR is needed, otherwise the output is too noisy for benchstat.

pohly committed:
scheduler_perf: add logging flags

This enables testing of different real production configurations (JSON
vs. text, different log levels, contextual logging).

pohly commented:
@kerthcet: I changed this PR so that it contains everything that is needed to generated output that "benchstat" can digest. We could merge it now.

Can you take another look?

kerthcet reviewed (commented):
Some comments, I'll experiment this tomorrow.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
+	//
+	// Because we leak one goroutine which calls klog, we cannot restore
+	// the previous state.
+	_ = framework.RedirectKlog(b, output)
```
"go test -race" reports a race when changing klog again while the leaked goroutine is running.

pohly commented on test/integration/scheduler_perf/main_test.go:
```diff
@@ -18,17 +18,55 @@ package benchmark
+	// logging flags. That's correct for most tests, but in the
+	// scheduler_perf test we want more control over the flags, therefore
+	// here strip them out.
+	var fs flag.FlagSet
```
This is a permanent special case for scheduler_perf. As far as I know, it is the only performance benchmark which uses the integration framework.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
+	output := initTestOutput(b)
+
+	// Because we run sequentially, it is possible to change the global
+	// klog logger and redirect log output. Quite a lot of code still uses
```
Hmm, this doesn't sound right to me. I don't think "source code" has a plural form - but I am not a native speaker.

pohly replied:
https://english.stackexchange.com/questions/584324/is-the-phrase-source-code-intrinsically-plural

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
+						// useful because it includes
+						// the time spent on
+						// initialization and shutdown. Here we suppress it.
+						b.ReportMetric(0, "ns/op")
```
And call that when?

Besides that, the "ns/op" metric is also not very useful. It highly depends on the configuration of the test case.

The new "average pods/s" is a better replacement for a measurement that covers the relevant duration of the measurement section of a test case.

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
+							// the actual value has some
+							// other unit. We patch the key
+							// to make it look right.
+							metric := strings.ReplaceAll(result.Labels["Metric"], "_seconds", "_"+result.Unit)
```
This reads from `result`, but doesn't modify it?

kerthcet replied:
I'll take a care of this. 
This PR makes benchmark results visible via benchstat, so
/lgtm

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
+					// quit *before* restoring klog settings.
+					framework.GoleakCheck(b)
+
+					ctx := context.Background()
```
This context is the parent of the timeout context. Depending on the parameters, it directly replaces the current `context.TODO()` (not using ktesting) or is a context which has the per-test logger (using ktesting).

pohly commented on test/integration/scheduler_perf/scheduler_perf_test.go:
```diff
@@ -570,6 +575,39 @@ func (so sleepOp) patchParams(_ *workload) (realOp, error) {
 	return &so, nil
 }
 
+var useTestingLog = flag.Bool("use-testing-log", false, "Write log entries with testing.TB.Log. This is more suitable for unit testing, but less realistic in real benchmarks.")
```
Added.

pohly commented on test/utils/format/format.go:
```diff
@@ -78,3 +78,8 @@ func useYAML(t reflect.Type) bool {
 	}
 }
+
+// AnyToString is a wrapper around format.Object with zero indention.
```
Removed.

kerthcet commented:
Besides, do you tidy up the output? Not the same as mine.

pohly commented:
You need to run `benchstat` on the output of `go test`. I've updated the PR description to show that.

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest
